Index: alsa-lib-1.0.18/src/conf/cards/CMI8788.conf
===================================================================
--- alsa-lib-1.0.18.orig/src/conf/cards/CMI8788.conf	2009-01-19 23:41:42.000000000 -0500
+++ alsa-lib-1.0.18/src/conf/cards/CMI8788.conf	2009-01-19 23:42:16.000000000 -0500
@@ -13,7 +13,7 @@
 	card $CARD
 }	
 
-# default with dmix & dsnoop
+# default with dmix+softvol & dsnoop
 CMI8788.pcm.default {
 	@args [ CARD ]
 	@args.CARD {
@@ -23,8 +23,15 @@
 	playback.pcm {
 		type plug
 		slave.pcm {
-			@func concat
-			strings [ "dmix:" $CARD ",FORMAT=S32_LE" ]
+			type softvol
+			slave.pcm {
+				@func concat
+				strings [ "dmix:" $CARD ",FORMAT=S32_LE" ]
+			}
+			control {
+				name "PCM Playback Volume"
+				card $CARD
+			}
 		}
 	}
 	capture.pcm {
