Index: alsa-lib-1.0.21a/src/pcm/pcm_direct.c
===================================================================
--- alsa-lib-1.0.21a.orig/src/pcm/pcm_direct.c	2009-12-12 13:50:47.000000000 -0500
+++ alsa-lib-1.0.21a/src/pcm/pcm_direct.c	2009-12-12 13:51:23.000000000 -0500
@@ -591,7 +591,10 @@
 
 int snd_pcm_direct_info(snd_pcm_t *pcm, snd_pcm_info_t * info)
 {
-	// snd_pcm_direct_t *dmix = pcm->private_data;
+	snd_pcm_direct_t *dmix = pcm->private_data;
+
+	if (dmix->spcm && !dmix->shmptr->use_server)
+		return snd_pcm_info(dmix->spcm, info);
 
 	memset(info, 0, sizeof(*info));
 	info->stream = pcm->stream;
