Index: alsa-lib-1.0.22/src/pcm/pcm_direct.c
===================================================================
--- alsa-lib-1.0.22.orig/src/pcm/pcm_direct.c	2010-03-03 08:19:38.000000000 -0500
+++ alsa-lib-1.0.22/src/pcm/pcm_direct.c	2010-03-03 08:19:52.000000000 -0500
@@ -540,7 +540,6 @@
 int snd_pcm_direct_timer_stop(snd_pcm_direct_t *dmix)
 {
 	snd_timer_stop(dmix->timer);
-	snd_pcm_direct_clear_timer_queue(dmix);
 	return 0;
 }
 
@@ -567,6 +566,7 @@
 	switch (snd_pcm_state(dmix->spcm)) {
 	case SND_PCM_STATE_XRUN:
 	case SND_PCM_STATE_SUSPENDED:
+	case SND_PCM_STATE_SETUP:
 		events |= POLLERR;
 		break;
 	default:
@@ -577,6 +577,7 @@
 			switch (snd_pcm_state(pcm)) {
 			case SND_PCM_STATE_XRUN:
 			case SND_PCM_STATE_SUSPENDED:
+			case SND_PCM_STATE_SETUP:
 				events |= POLLERR;
 				break;
 			default:
@@ -1125,8 +1126,9 @@
 	snd_timer_poll_descriptors(dmix->timer, &dmix->timer_fd, 1);
 	dmix->poll_fd = dmix->timer_fd.fd;
 
-	dmix->timer_event_suspend = 1<<SND_TIMER_EVENT_MSUSPEND;
-	dmix->timer_event_resume = 1<<SND_TIMER_EVENT_MRESUME;
+	dmix->timer_events = (1<<SND_TIMER_EVENT_MSUSPEND) |
+			     (1<<SND_TIMER_EVENT_MRESUME) |
+			     (1<<SND_TIMER_EVENT_STOP);
 
 	/*
 	 * Some hacks for older kernel drivers
@@ -1145,9 +1147,15 @@
 		 * suspend/resume events.
 		 */
 		if (ver < SNDRV_PROTOCOL_VERSION(2, 0, 5)) {
-			dmix->timer_event_suspend = 1<<SND_TIMER_EVENT_MPAUSE;
-			dmix->timer_event_resume = 1<<SND_TIMER_EVENT_MCONTINUE;
+			dmix->timer_events &= ~((1<<SND_TIMER_EVENT_MSUSPEND) |
+						(1<<SND_TIMER_EVENT_MRESUME));
+			dmix->timer_events |= (1<<SND_TIMER_EVENT_MPAUSE) |
+					      (1<<SND_TIMER_EVENT_MCONTINUE);
 		}
+		/* In older versions, use SND_TIMER_EVENT_START too.
+		 */
+		if (ver < SNDRV_PROTOCOL_VERSION(2, 0, 6))
+			dmix->timer_events |= 1<<SND_TIMER_EVENT_START;
 	}
 	return 0;
 }
@@ -1274,8 +1282,7 @@
 	snd_timer_params_set_ticks(params, 1);
 	if (dmix->tread) {
 		filter = (1<<SND_TIMER_EVENT_TICK) |
-			 dmix->timer_event_suspend |
-			 dmix->timer_event_resume;
+			 dmix->timer_events;
 		snd_timer_params_set_filter(params, filter);
 	}
 	ret = snd_timer_params(dmix->timer, params);
Index: alsa-lib-1.0.22/src/pcm/pcm_direct.h
===================================================================
--- alsa-lib-1.0.22.orig/src/pcm/pcm_direct.h	2010-03-03 08:19:38.000000000 -0500
+++ alsa-lib-1.0.22/src/pcm/pcm_direct.h	2010-03-03 08:19:52.000000000 -0500
@@ -142,10 +142,9 @@
 	int hw_fd;			/* hardware file descriptor */
 	struct pollfd timer_fd;
 	int poll_fd;
-	int tread;
-	int timer_need_poll;
-	unsigned int timer_event_suspend;
-	unsigned int timer_event_resume;
+	int tread: 1;
+	int timer_need_poll: 1;
+	unsigned int timer_events;
 	int server_fd;
 	pid_t server_pid;
 	snd_timer_t *timer; 		/* timer used as poll_fd */
Index: alsa-lib-1.0.22/src/pcm/pcm_dmix.c
===================================================================
--- alsa-lib-1.0.22.orig/src/pcm/pcm_dmix.c	2010-03-03 08:19:38.000000000 -0500
+++ alsa-lib-1.0.22/src/pcm/pcm_dmix.c	2010-03-03 08:19:52.000000000 -0500
@@ -597,8 +597,8 @@
 	snd_pcm_direct_t *dmix = pcm->private_data;
 	if (dmix->state == SND_PCM_STATE_OPEN)
 		return -EBADFD;
-	snd_pcm_direct_timer_stop(dmix);
 	dmix->state = SND_PCM_STATE_SETUP;
+	snd_pcm_direct_timer_stop(dmix);
 	return 0;
 }
 
Index: alsa-lib-1.0.22/src/pcm/pcm_dshare.c
===================================================================
--- alsa-lib-1.0.22.orig/src/pcm/pcm_dshare.c	2010-03-03 08:19:38.000000000 -0500
+++ alsa-lib-1.0.22/src/pcm/pcm_dshare.c	2010-03-03 08:19:52.000000000 -0500
@@ -355,9 +355,9 @@
 	snd_pcm_direct_t *dshare = pcm->private_data;
 	if (dshare->state == SND_PCM_STATE_OPEN)
 		return -EBADFD;
+	dshare->state = SND_PCM_STATE_SETUP;
 	snd_pcm_direct_timer_stop(dshare);
 	do_silence(pcm);
-	dshare->state = SND_PCM_STATE_SETUP;
 	return 0;
 }
 
Index: alsa-lib-1.0.22/src/pcm/pcm_dsnoop.c
===================================================================
--- alsa-lib-1.0.22.orig/src/pcm/pcm_dsnoop.c	2010-03-03 08:19:38.000000000 -0500
+++ alsa-lib-1.0.22/src/pcm/pcm_dsnoop.c	2010-03-03 08:19:52.000000000 -0500
@@ -280,8 +280,8 @@
 	snd_pcm_direct_t *dsnoop = pcm->private_data;
 	if (dsnoop->state == SND_PCM_STATE_OPEN)
 		return -EBADFD;
-	snd_timer_stop(dsnoop->timer);
 	dsnoop->state = SND_PCM_STATE_SETUP;
+	snd_timer_stop(dsnoop->timer);
 	return 0;
 }
 
