Index: alsa-lib-1.0.21a/src/mixer/simple.c
===================================================================
--- alsa-lib-1.0.21a.orig/src/mixer/simple.c	2009-12-12 14:16:52.000000000 -0500
+++ alsa-lib-1.0.21a/src/mixer/simple.c	2009-12-12 14:49:13.000000000 -0500
@@ -103,7 +103,7 @@
 }
 
 #define CHECK_ENUM(xelem) \
-	if (!((sm_selem_t *)(elem)->private_data)->caps & (SM_CAP_PENUM|SM_CAP_CENUM)) \
+	if (!(((sm_selem_t *)(elem)->private_data)->caps & (SM_CAP_PENUM|SM_CAP_CENUM))) \
 		return -EINVAL;
 
 #define COND_CAPS(xelem, what) \
@@ -878,7 +878,6 @@
 int snd_mixer_selem_is_enumerated(snd_mixer_elem_t *elem)
 {
 	CHECK_BASIC(elem);
-	CHECK_ENUM(elem);
 	return sm_selem_ops(elem)->is(elem, SM_PLAY, SM_OPS_IS_ENUMERATED, 0);
 }
 
