Index: alsa-lib-1.0.20/src/pcm/pcm.c
===================================================================
--- alsa-lib-1.0.20.orig/src/pcm/pcm.c	2009-09-27 16:37:14.000000000 -0400
+++ alsa-lib-1.0.20/src/pcm/pcm.c	2009-09-27 16:48:48.000000000 -0400
@@ -6406,7 +6406,7 @@
 	if (xareas == NULL)
 		return -EBADFD;
 	*areas = xareas;
-	*offset = *pcm->appl.ptr % pcm->buffer_size;
+	*offset = (pcm->buffer_size > 0) ? *pcm->appl.ptr % pcm->buffer_size : 0;
 	avail = snd_pcm_mmap_avail(pcm);
 	if (avail > pcm->buffer_size)
 		avail = pcm->buffer_size;
