Index: alsa-lib-1.0.18/include/sound/asound.h
===================================================================
--- alsa-lib-1.0.18.orig/include/sound/asound.h	2009-01-29 17:41:03.000000000 -0500
+++ alsa-lib-1.0.18/include/sound/asound.h	2009-01-29 17:41:35.000000000 -0500
@@ -402,7 +402,7 @@
 
 struct sndrv_pcm_channel_info {
 	unsigned int channel;
-	off_t offset;			/* mmap offset */
+	long offset;			/* mmap offset */
 	unsigned int first;		/* offset to first sample in bits */
 	unsigned int step;		/* samples distance in bits */
 };
