Index: alsa-lib-1.0.18/src/pcm/pcm_plug.c
===================================================================
--- alsa-lib-1.0.18.orig/src/pcm/pcm_plug.c	2009-02-12 18:19:31.000000000 -0500
+++ alsa-lib-1.0.18/src/pcm/pcm_plug.c	2009-02-12 18:19:50.000000000 -0500
@@ -612,6 +612,8 @@
 	case SND_PCM_ACCESS_RW_NONINTERLEAVED:
 		slv->access = SND_PCM_ACCESS_MMAP_NONINTERLEAVED;
 		break;
+	default:
+		break;
 	}
 	return 1;
 }
