Index: alsa-lib-1.0.18/src/pcm/pcm_meter.c
===================================================================
--- alsa-lib-1.0.18.orig/src/pcm/pcm_meter.c	2009-01-14 18:37:52.000000000 -0500
+++ alsa-lib-1.0.18/src/pcm/pcm_meter.c	2009-01-14 18:38:11.000000000 -0500
@@ -603,7 +603,7 @@
 	snd_config_iterator_t i, next;
 	const char *id;
 	const char *lib = NULL, *open_name = NULL, *str = NULL;
-	snd_config_t *c, *type_conf;
+	snd_config_t *c, *type_conf = NULL;
 	int (*open_func)(snd_pcm_t *, const char *,
 			 snd_config_t *, snd_config_t *) = NULL;
 	snd_pcm_meter_t *meter = pcm->private_data;
