Index: alsa-lib-1.0.18/src/conf.c
===================================================================
--- alsa-lib-1.0.18.orig/src/conf.c	2009-01-14 18:42:04.000000000 -0500
+++ alsa-lib-1.0.18/src/conf.c	2009-01-14 18:42:14.000000000 -0500
@@ -2962,7 +2962,7 @@
 	configs = cfgs;
 	if (!configs) {
 		configs = getenv(ALSA_CONFIG_PATH_VAR);
-		if (!configs)
+		if (!configs || !*configs)
 			configs = ALSA_CONFIG_PATH_DEFAULT;
 	}
 	for (k = 0, c = configs; (l = strcspn(c, ": ")) > 0; ) {
