Index: alsa-lib-1.0.23/src/control/namehint.c
===================================================================
--- alsa-lib-1.0.23.orig/src/control/namehint.c	2010-10-03 14:53:13.000000000 -0400
+++ alsa-lib-1.0.23/src/control/namehint.c	2010-10-03 14:56:02.000000000 -0400
@@ -439,14 +439,15 @@
 			err = next_devices[list->iface](list->ctl, &device);
 			if (device < 0)
 				err = -EINVAL;
+			else
+				max_device = device;
 			while (err >= 0 && device >= 0) {
 				err = next_devices[list->iface](list->ctl, &device);
-				if (device > max_device)
+				if (err >= 0 && device > max_device)
 					max_device = device;
-				ok++;
 			}
 			ok = 0;
-			for (device = 0; err >= 0 && device < max_device; device++) {
+			for (device = 0; err >= 0 && device <= max_device; device++) {
 				list->device = device;
 				err = try_config(list, list->siface, str);
 				if (err < 0)
