diff -urN alsa-lib-1.0.17a/src/conf.c alsa-lib-1.0.17a.new/src/conf.c
--- alsa-lib-1.0.17a/src/conf.c	2008-08-15 21:28:55.000000000 +1000
+++ alsa-lib-1.0.17a.new/src/conf.c	2008-09-30 13:19:56.000000000 +1000
@@ -2607,7 +2607,9 @@
 #define ALSA_CONFIG_PATH_VAR "ALSA_CONFIG_PATH"
 
 /** The name of the default files used by #snd_config_update. */
-#define ALSA_CONFIG_PATH_DEFAULT ALSA_CONFIG_DIR "/alsa.conf"
+#ifndef ALSA_CONFIG_PATH_DEFAULT
+  #define ALSA_CONFIG_PATH_DEFAULT ALSA_CONFIG_DIR "/alsa.conf"
+#endif
 
 /**
  * \ingroup Config
