Index: alsa-lib-1.0.18/src/pcm/pcm.c
===================================================================
--- alsa-lib-1.0.18.orig/src/pcm/pcm.c	2009-02-12 18:21:40.000000000 -0500
+++ alsa-lib-1.0.18/src/pcm/pcm.c	2009-02-12 18:21:54.000000000 -0500
@@ -7257,7 +7257,7 @@
  * \param channels required PCM channels
  * \param rate required sample rate in Hz
  * \param soft_resample 0 = disallow alsa-lib resample stream, 1 = allow resampling
- * \param latency required overall latency in us (0 = optimum latency for players)
+ * \param latency required overall latency in us
  * \return 0 on success otherwise a negative error code
  */
 int snd_pcm_set_params(snd_pcm_t *pcm,
