Description: Explicitly include <sys/types.h>
Author: Adam Conrad <adconrad@ubuntu.com>
Bug-Ubuntu: https://bugs.launchpad.net/bugs/1109298
Forwarded: no

--- alsa-lib-1.0.25.orig/include/asoundlib.h
+++ alsa-lib-1.0.25/include/asoundlib.h
@@ -31,6 +31,7 @@
 #include <unistd.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <sys/types.h>
 #include <string.h>
 #include <fcntl.h>
 #include <assert.h>
